css-transitions-2/Overview.bs

```diff
@@ -169,23 +169,38 @@ transition's <dfn>transition generation</dfn>.
 
 ## Defining [=before-change style=]: the ''@starting-style'' rule
 
-Transitions can only start during a [=style change event=] for elements which
-has a defined [=before-change style=] established by the previous
-[=style change event=], which means that newly inserted elements or elements
-that change from not [=being rendered=] to being rendered do not start
-transitions. Still, doing transitions starting from the initial rendering is a
-valid use case.
+In Level 1 of this specification, transitions can only start during a
+[=style change event=] for elements which have a defined [=before-change style=]
+established by the previous [=style change event=]. That means a transition
+could not be started on an element that was not being rendered for the previous
+[=style change event=] (see: [[css-transitions-1#starting]]).
+
+In some cases it makes sense to start transitions on newly inserted elements or
+elements that change from not [=being rendered=] to being rendered. To allow
+for that, this specification introduces ''@starting-style''.
 
 The <dfn at-rule id="at-ruledef-starting-style">@starting-style</dfn> rule is a
-conditional group rule with the implicit condition that it matches if the
-previous [=style change event=] did not establish a [=before-change style=] for
-the element whose styles are being computed. This is used to establish a
+grouping rule where the rules inside it apply if the previous
+[=style change event=] did not establish a [=before-change style=] for the
+element whose styles are being computed. This is used to establish a
 [=before-change style=] for the current [=style change event=].
 
-If any rules in ''@starting-style'' are applied, the resulting style must be
-used as the [=before-change style=] for the current [=style change event=] and
-the computed styles must be re-computed without ''@starting-style'' rules
-applied.
+Note that this means that ''@starting-style'' rules only applies to some
+elements during a computed style update, namely elements that were not rendered
+or part of the DOM during the previous [=style change event=].
+
+If any style rules in ''@starting-style'' are applied, the resulting style must
+be used as the [=before-change style=] for the current [=style change event=].
+Style rules in ''@starting-style'' do not apply directly to
+[=after-change style=].
+
+<div class="note">
+Since ''@starting-style'' is applied to create a [=before-change style=] during
+the same [=style change event=] as it is being used for triggering
+transitions, it inherits from the parent element's [=after-change style=], even
```

Comment:
I don't think before-change style inheriting from parent's after-change style is correct per the current css-transitions-1 spec... although it's possible there are unlanded changes that @andruud was working on that would make it correct??  (But I think the historic change was perhaps in the other direction?)

Reply:
This is about computing before-change style in the style change event as the after-change style, which is a new thing, so I don't understand how that would strictly be in conflict with existing specs?

You would not be able to completely create the before-change style as of the previous style change event as the state of the elements involved has already changed (for selector matching, for instance).

Reply:
OK... I agree it is somewhat different.  That said, I think there were good reasons (related to container queries) for the setup that I *think* is reflected in the current spec, and I don't completely remember what they are, and I hope @andruud does.

I also don't understand what would be hard about creating the before-change style for the parent -- presumably if the parent had needed to start a transition, it would have used that before-change style, so it should still be around.  Or am I missing something here?

Reply:
That would generally require two computed style trees to be kept alive, the new and the old one. This is an implementation detail, but what happens in Blink for container queries is that we temporarily store the before-change style on the first resolve if we risk doing multiple style passes because of multiple layout passes for a query container so that we use the original before-change style and not update the before-change style for every layout pass. However, we only do that when inside a size container and the style may be affected by animations/transitions.

If @starting-style should inherit from before-change style from the previous style-change if it existed, consider the following case:

```html
<div id="a">
  <div id="b">
    <div id="c"></div>
  </div>
</div>
```

Say that #b was display:none in the previous style-change, and we have @starting-style rules for #c. The before-change style for #a is available from the previous style-change, but we have no idea what the style for #b would have been if it was created and stored for the previous style-change event.

Then, if the before-change style for #c should inherit the before-change style from #b, we have no better idea than to use the after-change style for #b as any state for #b may have changed since the last style-change event. We could inherit before-change style from #a and create two separate styles for #b, one before-change style, in case some descendant needs it to inherit from when starting transitions, and one for the after-change style. 

Now, I think that is a pretty big extra burden on an implementation, and still while inheritance may make the before-change style include more styles from the previous style-change event, it still will be heavily influenced by the current style-change event due to selectors matching the current DOM state.

Reply:
The unlanded change that I never managed to follow up on, stated roughly:

 - The _base style_ is the style as it would be without animations/transitions/SMIL, inheriting from the base style of the parent.
 - The before/after-change style is just the base style observed at different points in time. Hence "cross-inheritance" between before/after-change styles is indeed not a thing in this model.

The reason for proposing that model is that the current spec (if I recall correctly) defines the before-change style to be the before-change base style _plus_ animations/transitions/etc updated to the _current time_. Say we were animating a width for example, and had some CQs (ultimately) affected by that width. At the moment we're calculating the before-change style, the "change" has already happened. The entire DOM could have been mutated heavily. We are no longer able to evaluate the CQs in a world where that DOM mutation _didn't_ happen, because we don't keep around multiple worlds.

This was already a hard problem pre-CQ, because the spec required to keep around multiple "responsive" style worlds: e.g. a font-size animation would affect any em units found in the before-change (base) style (Blink never implemented this). With CQ this extends to all of layout, making it probably too hard to be ever worth considering.

---

@lilles' before-change style inheriting from an after-change style seemed really weird to me at first as well. But he had a good point that the starting-style is produced matching against a current (i.e. after-change) DOM.

Possibly we should avoid saying that @starting-style affects the before-style change itself, and instead spec that:

 - The _starting style_ is the before-change style if it exsts, otherwise produced from @starting-style and inheriting from whatever makes sense.
 - A transition is between the starting style and the after-change style.

Reply:
So I think the *intent* of the current model is that the updating of the before-change style doesn't involve any changes to rule matching -- it just involves taking the already-matched animations/transitions/etc. and updating the time-based parts of them to the current time.  Maybe that isn't stated explicitly, but it's possibly it simplifies the CQ problem.  Though I'm not sure if it's relevant here.

It is indeed true that in the case where the before-change style comes from `@starting-style` rather than from a previous style change event, it is based on matching against the current DOM rather than a previous DOM.  But I think that's fine.

Perhaps the spec should be clearer about this, but I don't think it's possible for an element to both (a) have a before-change style from a previous style change event *and* (b) have its after-change style inherit from a different element than its before-change style.  This is because (and I'm not sure the spec is clear about this) a change in position in the DOM is effectively a removal and re-addition to the document, which is equivalent to not being rendered in the previous style change event.

Given those two things (which you should feel free to disagree with if you think they're wrong -- although I do think they're correct), it's not clear to me why the argument about changes in the DOM is relevant to this question.  I think it's fundamentally about what should happen to the before-change style (in particular, what it inherits from) at the boundary where an element that was previously not being rendered is the child of an element that *was* previously rendered.

It's not clear to me whether introducing a new "starting style" term would help.  If introducing a new term makes things clearer, then doing so seems fine.  But I think the main thing I'm asking about here is what the correct behavior is rather than about how to explain it.

Reply:
The CQ problem is about multiple passes of computing styles happening during the same style change event because we can have multiple layout passes for a given container with different size for each pass, but for transitions it is an atomic event. Otherwise the number of passes for the layout engine would be observable through transitions. I don't think that is relevant for @starting-style as such.

I think @andruud's point about matching the current DOM was about attribute and state changes, not about parent/child mutations.

Reply:
OK, given the explanation in https://github.com/w3c/csswg-drafts/pull/8876#discussion_r1236644072 I'm OK with this approach.